packages/block-editor/src/components/use-setting/index.js

```diff
@@ -139,29 +142,39 @@ export default function useSetting( path ) {
 						false
 					)
 				) {
-					const candidateAtts =
+					const attributes =
 						select( blockEditorStore ).getBlockAttributes(
 							candidateClientId
 						);
-					const candidateResult =
-						get(
-							candidateAtts,
-							`settings.blocks.${ blockName }.${ normalizedPath }`
-						) ??
-						get( candidateAtts, `settings.${ normalizedPath }` );
-					if ( candidateResult !== undefined ) {
-						result = candidateResult;
+					result =
+						get( attributes, [
+							'settings',
+							'blocks',
+							blockName,
+							normalizedPath,
+						] ) ??
+						get( attributes, [ 'settings', normalizedPath ] );
+					if ( result !== undefined ) {
+						// Stop the search for more distant ancestors and move on.
+						break;
 					}
 				}
-			} );
+			}
 
 			// 2. Fall back to the settings from the block editor store (__experimentalFeatures).
 			const settings = select( blockEditorStore ).getSettings();
 			if ( result === undefined ) {
-				const defaultsPath = `__experimentalFeatures.${ normalizedPath }`;
-				const blockPath = `__experimentalFeatures.blocks.${ blockName }.${ normalizedPath }`;
 				result =
-					get( settings, blockPath ) ?? get( settings, defaultsPath );
+					get( settings, [
```

Comment:
To use the array syntax with `get` you need to "normalize" the normalized path first (it may or may not have a dot syntax, such as `color.text` or `layout` ).

Reply:
Ah, that makes all the sense in the world, and I clearly got carried away in the refactor. :) 

Just for fun, I confirmed that this works:

```js
get(
  settings,
  path.flatMap( branch => branch.split( '.' ) ),
  default
)
```

but, given the (obvious) expectation around `normalizedPath`, it's best to just revert that particular commit.